File: Samples/BarcodeScanner/js/js/scenario3_ActiveSymbologies.js
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

(function () {
    "use strict";

    // Types
    var BarcodeSymbologies = Windows.Devices.PointOfService.BarcodeSymbologies;

    // Page elements
    var scenarioStartScanButton;
    var scenarioEndScanButton;
    var setActiveSymbologiesButton;
    var scenarioOutputScanDataType;
    var scenarioOutputScanData;
    var scenarioOutputScanDataFormatted;

    // WinJS controls
    var symbologyListViewControl = null;

    var page = WinJS.UI.Pages.define("/html/scenario3_ActiveSymbologies.html", {

        // On ready - set button states and add event listeners to the buttons.
        ready: function (element, options) {

            WinJS.log("Click the start scanning button to begin.", "sample", "status");

            scenarioOutputScanDataType = document.getElementById("scenarioOutputScanDataType");
            scenarioOutputScanData = document.getElementById("scenarioOutputScanData");
            scenarioOutputScanDataFormatted = document.getElementById("scenarioOutputScanDataFormatted");

            scenarioStartScanButton = document.getElementById("scenarioStartScanButton");
            scenarioStartScanButton.addEventListener("click", startReceivingData, false);
            scenarioStartScanButton.disabled = false;

            scenarioEndScanButton = document.getElementById("scenarioEndScanButton");
            scenarioEndScanButton.addEventListener("click", endReceivingData, false);
            scenarioEndScanButton.disabled = true;

            setActiveSymbologiesButton = document.getElementById("setActiveSymbologiesButton");
            setActiveSymbologiesButton.addEventListener("click", setActiveSymbologies, false);
            setActiveSymbologiesButton.disabled = true;

            symbologyListViewControl = document.getElementById("symbologyListView").winControl;
        },

        unload: function () {
            closeScanners();
        }
    });

    var _scanner = null;
    var _claimedScanner = null;

    // Creates the default barcode scanner and claims it. Once, claimed, it adds event listeners for onDataReceived and releaseDeviceRequested events.
    // On error, it logs the error message.
    function startReceivingData() {
        scenarioStartScanButton.disabled = true;

        // Create the barcode scanner.
        WinJS.log("Creating barcode scanner object.", "sample", "status");

        SdkSample.getFirstBarcodeScannerAsync().then(function (scanner) {
            _scanner = scanner;
            if (scanner) {
                // After successful creation, fill the list with supported symbologies.
                return _scanner.getSupportedSymbologiesAsync().then(function (symbologies) {
                    var listOfSymbologies = new WinJS.Binding.List(symbologies.map(function (symbology) {
                        return { id: symbology, name: BarcodeSymbologies.getName(symbology) };
                    }));
                    symbologyListViewControl.itemDataSource = listOfSymbologies.dataSource;

                    // Claim the scanner for exclusive use.
                    return _scanner.claimScannerAsync();
                });
            } else {
                WinJS.log("Barcode scanner not found. Please connect a barcode scanner.", "sample", "error");
            }
        }).then(function (claimedScanner) {
            _claimedScanner = claimedScanner;
            if (claimedScanner) {

                // Ask the API to decode the data by default. By setting this, API will decode the raw data from the barcode scanner and 
                // send the ScanDataLabel and ScanDataType in the DataReceived event
                claimedScanner.isDecodeDataEnabled = true;

                //// After successfully claiming, attach the datareceived event handler.
                claimedScanner.addEventListener("datareceived", onDataReceived);

                // It is always a good idea to have a release device requested event handler. If this event is not handled, there are chances of another app can 
                // claim ownership of the barcode scanner.
                claimedScanner.addEventListener("releasedevicerequested", onReleasedeviceRequested);

                // Enable the scanner so it starts raising events.
                return claimedScanner.enableAsync();
            } else if (_scanner) {
                _scanner.close();
                _scanner = null;
                WinJS.log("Claim barcode scanner failed.", "sample", "error");
            }
        }).done(function () {
            if (_claimedScanner) {
                WinJS.log("Ready to scan. Device ID: " + _claimedScanner.deviceId, "sample", "status");
                scenarioEndScanButton.disabled = false;
                setActiveSymbologiesButton.disabled = false;
            } else {
                scenarioStartScanButton.disabled = false;
            }
        });
    }

    // Event handler for the Release Device Requested event fired when barcode scanner receives Claim request from another application
    function onReleasedeviceRequested(args) {
        _claimedScanner.retainDevice();
        WinJS.log("Event ReleaseDeviceRequested received. Retaining the barcode scanner.", "sample", "status");
    }

    // Event handler for the DataReceived event fired when a barcode is scanned by the barcode scanner 
    function onDataReceived(args) {
        scenarioOutputScanDataType.textContent = BarcodeSymbologies.getName(args.report.scanDataType);
        scenarioOutputScanData.textContent = getDataString(args.report.scanData);
        scenarioOutputScanDataFormatted.textContent = getDataLabelString(args.report.scanDataLabel, args.report.scanDataType);
    }

    // This method removes the event listeners from the claimed scanner object if it is not null and then closes the claimed scanner so that someone else can use it.
    function endReceivingData() {
        closeScanners();

        // Reset button states
        WinJS.log("Click the start scanning button to begin.", "sample", "status");

        setActiveSymbologiesButton.disabled = true;
        scenarioStartScanButton.disabled = false;
        scenarioEndScanButton.disabled = true;
        scenarioOutputScanDataType.textContent = "No Data";
        scenarioOutputScanData.textContent = "No Data";
        scenarioOutputScanDataFormatted.textContent = "No Data";

        // reset symbology list
        symbologyListViewControl.itemDataSource = null;
    }

    // This method sets the active symbologies of the claimed scanner object.
    function setActiveSymbologies() {
        if (_claimedScanner) {

            // Collect the symbology IDs the user selected in the list control.
            symbologyListViewControl.selection.getItems().done(function (items) {
                var activeSymbologies = items.map(function (item) {
                    return item.data.id;
                })
                // Set them as active in the claimed scanner.
                _claimedScanner.setActiveSymbologiesAsync(activeSymbologies);
            });
        }
    }

    function closeScanners() {
        if (_claimedScanner) {
            _claimedScanner.removeEventListener("datareceived", onDataReceived);
            _claimedScanner.removeEventListener("releasedevicerequested", onReleasedeviceRequested);
            _claimedScanner.close();
            _claimedScanner = null;
        }
        if (_scanner) {
            _scanner.close();
            _scanner = null;
        }
    }
})();

File: Samples/BarcodeScanner/cppwinrt/Scenario3_ActiveSymbologies.cpp
#include "pch.h"
#include "Scenario3_ActiveSymbologies.h"
#include "Scenario3_ActiveSymbologies.g.cpp"

namespace winrt
{
    using namespace Windows::Foundation;
    using namespace Windows::Foundation::Collections;
    using namespace Windows::Devices::PointOfService;
    using namespace Windows::UI::Core;
    using namespace Windows::UI::Xaml;
    using namespace Windows::UI::Xaml::Data;
    using namespace Windows::UI::Xaml::Controls;
    using namespace Windows::UI::Xaml::Navigation;
}

namespace winrt::SDKTemplate::implementation
{
    Scenario3_ActiveSymbologies::Scenario3_ActiveSymbologies()
    {
        InitializeComponent();

        m_listOfSymbologies = single_threaded_observable_vector<SDKTemplate::SymbologyListEntry>();
        SymbologyListSource().Source(m_listOfSymbologies);
    }

    /// <summary>
    /// Invoked when this page is about to be displayed in a Frame.
    /// </summary>
    void Scenario3_ActiveSymbologies::OnNavigatedTo(NavigationEventArgs const&)
    {
        ResetTheScenarioState();
    }

    /// <summary>
    /// Invoked when this page is no longer displayed.
    /// </summary>
    void Scenario3_ActiveSymbologies::OnNavigatedFrom(NavigationEventArgs const&)
    {
        ResetTheScenarioState();
    }

    /// <summary>
    /// Event Handler for Start Scan Button Click.
    /// Sets up the barcode scanner to be ready to receive the data events from the scan.
    /// </summary>
    fire_and_forget Scenario3_ActiveSymbologies::ScenarioStartScanButton_Click(IInspectable const&, RoutedEventArgs const&)
    {
        auto lifetime = get_strong();

        ScenarioStartScanButton().IsEnabled(false);

        m_rootPage.NotifyUser(L"Acquiring barcode scanner object.", NotifyType::StatusMessage);

        // create the barcode scanner.
        m_scanner = co_await BarcodeScanner::GetDefaultAsync();

        if (m_scanner)
        {
            // Claim the scanner for exclusive use and enable it so raises DataReceived events.
            m_claimedScanner = co_await m_scanner.ClaimScannerAsync();
            if (m_claimedScanner)
            {
                // It is always a good idea to have a release device requested event handler.
                // If this event is not handled, then another app can claim ownership of the barcode scanner.
                m_scannerReleaseRequestedToken = m_claimedScanner.ReleaseDeviceRequested({ get_weak(), &Scenario3_ActiveSymbologies::claimedScanner_ReleaseDeviceRequested });

                // after successfully claiming, attach the datareceived event handler.
                m_scannerDataReceivedToken = m_claimedScanner.DataReceived({ get_weak(), &Scenario3_ActiveSymbologies::claimedScanner_DataReceived });

                // Ask the platform to decode the data by default. When this is set, the platform
                // will decode the raw data from the barcode scanner and include in the
                // BarcodeScannerReport.ScanDataLabel and ScanDataType in the DataReceived event.
                m_claimedScanner.IsDecodeDataEnabled(true);

                // Enable the scanner so it raises DataReceived events.
                // Do this after adding the DataReceived event handler.
                co_await m_claimedScanner.EnableAsync();

                // after successful claim, list supported symbologies
                auto supportedSymbologies = co_await m_scanner.GetSupportedSymbologiesAsync();
                for (uint32_t symbology : supportedSymbologies)
                {
                    m_listOfSymbologies.Append(SymbologyListEntry{ symbology });
                }

                // reset the button state
                ScenarioEndScanButton().IsEnabled(true);
                SetActiveSymbologiesButton().IsEnabled(true);

                m_rootPage.NotifyUser(L"Ready to scan. Device ID: " + m_claimedScanner.DeviceId(), NotifyType::StatusMessage);
            }
            else
            {
                m_scanner.Close();
                m_scanner = nullptr;
                ScenarioStartScanButton().IsEnabled(true);
                m_rootPage.NotifyUser(L"Claim barcode scanner failed.", NotifyType::ErrorMessage);
            }
        }
        else
        {
            ScenarioStartScanButton().IsEnabled(true);
            m_rootPage.NotifyUser(L"Barcode scanner not found. Please connect a barcode scanner.", NotifyType::ErrorMessage);
        }
    }

    /// <summary>
    /// Event handler for the Release Device Requested event fired when barcode scanner receives Claim request from another application
    /// </summary>
    /// <param name="e"> Contains the ClamiedBarcodeScanner that is sending this request</param>
    void Scenario3_ActiveSymbologies::claimedScanner_ReleaseDeviceRequested(IInspectable const&, ClaimedBarcodeScanner const& e)
    {
        // always retain the device
        e.RetainDevice();
        m_rootPage.NotifyUser(L"Event ReleaseDeviceRequested received. Retaining the barcode scanner.", NotifyType::StatusMessage);
    }


    /// <summary>
    /// Event handler for the DataReceived event fired when a barcode is scanned by the barcode scanner
    /// </summary>
    /// <param name="args"> Contains the BarcodeScannerDataReceivedEventArgs which contains the data obtained in the scan</param>
    fire_and_forget Scenario3_ActiveSymbologies::claimedScanner_DataReceived(ClaimedBarcodeScanner const&, BarcodeScannerDataReceivedEventArgs args)
    {
        auto lifetime = get_strong();

        // need to update the UI data on the dispatcher thread.
        // update the UI with the data received from the scan.
        co_await resume_foreground(Dispatcher());

        // read the data from the buffer and convert to string.
        ScenarioOutputScanDataLabel().Text(GetDataLabelString(args.Report().ScanDataLabel(), args.Report().ScanDataType()));
        ScenarioOutputScanData().Text(GetDataString(args.Report().ScanData()));
        ScenarioOutputScanDataType().Text(BarcodeSymbologies::GetName(args.Report().ScanDataType()));
    }

    /// <summary>
    /// Reset the Scenario state
    /// </summary>
    void Scenario3_ActiveSymbologies::ResetTheScenarioState()
    {
        if (m_claimedScanner)
        {
            // Detach the event handlers
            m_claimedScanner.DataReceived(m_scannerDataReceivedToken);
            m_claimedScanner.ReleaseDeviceRequested(m_scannerReleaseRequestedToken);

            // Release the Barcode Scanner and set to nullptr
            m_claimedScanner.Close();
            m_claimedScanner = nullptr;
        }

        if (m_scanner)
        {
            m_scanner.Close();
            m_scanner = nullptr;
        }

        // Reset the UI if we are still the current page.
        if (Frame().Content() == *this)
        {
            m_rootPage.NotifyUser(L"Click the start scanning button to begin.", NotifyType::StatusMessage);
            ScenarioOutputScanData().Text(L"No data");
            ScenarioOutputScanDataLabel().Text(L"No data");
            ScenarioOutputScanDataType().Text(L"No data");

            // reset the button state
            SetActiveSymbologiesButton().IsEnabled(false);
            ScenarioEndScanButton().IsEnabled(false);
            ScenarioStartScanButton().IsEnabled(true);

            // reset symbology list
            m_listOfSymbologies.Clear();
        }
    }

    /// <summary>
    /// Event handler for End Scan Button Click.
    /// Releases the Barcode Scanner and resets the text in the UI
    /// </summary>
    void Scenario3_ActiveSymbologies::ScenarioEndScanButton_Click(IInspectable const&, RoutedEventArgs const&)
    {
        ResetTheScenarioState();
    }

    fire_and_forget Scenario3_ActiveSymbologies::SetActiveSymbologies_Click(IInspectable const&, RoutedEventArgs const&)
    {
        auto lifetime = get_strong();

        if (m_claimedScanner)
        {
            std::vector<uint32_t> symbologyList;
            for (auto&& symbologyListEntry : m_listOfSymbologies)
            {
                if (symbologyListEntry.IsEnabled())
                {
                    symbologyList.push_back(symbologyListEntry.Id());
                }
            }

            co_await m_claimedScanner.SetActiveSymbologiesAsync(std::move(symbologyList));
        }
    }
}